src/main/java/com/faforever/client/player/PlayerService.java

```diff
@@ -139,51 +144,54 @@ private void resetIdleTime(Player playerForUsername) {
     Optional.ofNullable(playerForUsername).ifPresent(player -> player.setIdleSince(Instant.now()));
   }
 
-  private void updateGamePlayers(List<String> players, Game game) {
-    players.stream()
+  private void updateGamePlayers(List<String> currentPlayers, Game game) {
+    log.info(currentPlayers.toString());
+    currentPlayers.stream()
         .map(this::getPlayerForUsername)
         .filter(Optional::isPresent)
         .map(Optional::get)
         .forEach(player -> {
           resetIdleTime(player);
-          updateGameToPlayer(game, player);
-          if (game == null || game.getStatus() == GameStatus.CLOSED) {
-            player.setGame(null);
-          } else {
-            player.setGame(game);
-            if ((player.getGame() == null || !player.getGame().equals(game)) && player.getSocialStatus() == FRIEND && game.getStatus() == GameStatus.OPEN) {
-              eventBus.post(new FriendJoinedGameEvent(player, game));
-            }
-          }
+          updateGameDataForPlayer(game, player);
         });
 
+    //We need to see if anybody dropped out of games
     if (game != null && game.getStatus() != GameStatus.CLOSED && playersByGame.get(game.getId()) != null) {
-      playersByGame.get(game.getId()).stream()
-          .forEach(player -> {
-            boolean stillInGame = game.getTeams().entrySet().stream()
-                .anyMatch(entry -> entry.getValue().contains(player.getUsername()));
-            if (!stillInGame) {
-              player.setGame(null);
-            }
-          });
+      List<Player> playersThatLeftTheGame = new ArrayList<>();
```

Comment:
There where somethings going wrong here which caused the Friend joined game message not to be shown. I am not too sure what it was in particular, becuse it was more than one thing and I just cleaned everything up a little.

Also I added unit tests to make sure the new implementation is working